src/dos/program_tree.cpp

```diff
@@ -0,0 +1,446 @@
+		// draw the tree, use standard 7-bit ASCII characters
+
+		std::string tmp_str;
+		UTF8_RenderForDos("─├│└", tmp_str, code_page);
+		if (tmp_str != "\xC4\xC3\xB3\xC0") {
+			use_ascii_fallback = true;
+		}
+	}
+
+	if (has_option_ascii || use_ascii_fallback) {
+		str_child  = "|---";
+		str_last   = "\\---";
+		str_indent = "|   ";
+	} else {
+		UTF8_RenderForDos("├───", str_child, code_page);
+		UTF8_RenderForDos("└───", str_last, code_page);
+		UTF8_RenderForDos("│   ", str_indent, code_page);
+	}
+}
+
+void TREE::MaybeDisplayInfo(MoreOutputStrings& output, const DOS_DTA::Result& entry)
+{
+	if (has_option_size && has_option_files) {
+		if (entry.IsFile()) {
+			output.AddString("%13s ", format_number(entry.size).c_str());
+		} else {
+			output.AddString("              ");
+		}
+		if (has_option_attr) {
+			output.AddString("  ");
+		}
+	}
+	if (has_option_attr) {
+		output.AddString("%c %c%c%c ",
+		                 entry.attr.archive ? 'A' : '-',
+		                 entry.attr.hidden ? 'H' : '-',
+		                 entry.attr.system ? 'S' : '-',
+		                 entry.attr.read_only ? 'R' : '-');
+	}
+}
+
+void TREE::MaybeDisplayInfoSpace(MoreOutputStrings& output)
+{
+	if (has_option_size && has_option_files) {
+		output.AddString("              ");
+		if (has_option_attr) {
+			output.AddString("  ");
+		}
+	}
+	if (has_option_attr) {
+		output.AddString("      ");
+	}
+}
+
+uint8_t TREE::GetInfoSpaceSize() const
+{
+	uint8_t result = 0;
+	if (has_option_size && has_option_files) {
+		result = static_cast<uint8_t>(result + 14);
+		if (has_option_attr) {
+			result = static_cast<uint8_t>(result + 2);
+		}
+	}
+	if (has_option_attr) {
+		result = static_cast<uint8_t>(result + 6);
+	}
+	return result;
+}
+
+bool TREE::DisplayTree(MoreOutputStrings& output, const std::string& path,
+                       const uint16_t depth, const std::string& tree)
+{
+	auto display_empty = [&] {
+		if (skip_empty_line) {
+			return;
+		}
+		MaybeDisplayInfoSpace(output);
+		output.AddString("%s\n", tree.c_str());
+		skip_empty_line = true;
+	};
+
+	auto should_display = [&](const DOS_DTA::Result& result) {
+		if (!has_option_hidden && (result.attr.system | result.attr.hidden)) {
+			return false;
+		} else if (!result.IsDirectory()) {
+			return has_option_files;
+		} else if (result.IsDummyDirectory()) {
+			return false;
+		}
+		return true;
+	};
+
+	std::vector<DOS_DTA::Result> dir_contents;
+	size_t num_subdirs = 0;
+
+	// Get directory content
+
+	RealPt save_dta = dos.dta();
+	dos.dta(dos.tables.tempdta);
+
+	const auto pattern = path + "*.*";
+	FatAttributeFlags flags;
+	flags.system        = true;
+	flags.hidden        = true;
+	flags.directory     = true;
+	bool has_next_entry = DOS_FindFirst(pattern.c_str(), flags.data);
+	size_t space_needed = 7; // length of indentation + ellipsis
+
+	while (!shutdown_requested && has_next_entry) {
+		DOS_DTA::Result result = {};
+
+		DOS_DTA dta(dos.dta());
```

Comment:
can be `const`

Reply:
OK, changed - also in several other places